### ISSUE TITLE
Segfault passing b'\xff' into np.dtype (array underflow)

### ISSUE THREAD

**Issue by stestagg** [numpy/numpy#13902]
### Reproduction

```
>>> import numpy
>>> numpy.dtype(b'\xff')
Segmentation fault
```

### Error message:

Segfault
```
Program received signal SIGSEGV, Segmentation fault.                                                                                                                      
PyArray_DescrFromType (type=type@entry=-1) at numpy/core/src/multiarray/arraytypes.c.src:4468                                                                             
4468            Py_INCREF(ret);                                                                                                                                           
(gdb) bt                                                                                                                                                                  
#0  PyArray_DescrFromType (type=type@entry=-1) at numpy/core/src/multiarray/arraytypes.c.src:4468                                                                         
#1  0x00007f4d51f9445d in PyArray_DescrConverter (obj=0x7f4d52357a80, at=at@entry=0x7ffc13e5aa50) at numpy/core/src/multiarray/descriptor.c:1635                          
#2  0x00007f4d51f962bf in arraydescr_new (__NPY_UNUSED_TAGGEDsubtype=<optimized out>, args=<optimized out>, kwds=<optimized out>)                                         
    at numpy/core/src/multiarray/descriptor.c:2257                                                                                                                        
#3  0x00000000005d8e67 in _PyObject_FastCallKeywords ()                                                                                                                   
#4  0x000000000054aec1 in ?? ()                                                                                                                                           
#5  0x0000000000551f0a in _PyEval_EvalFrameDefault ()                                                                                                                     
#6  0x000000000054b7c2 in _PyEval_EvalCodeWithName ()                                                                                                                     
#7  0x000000000054dae3 in PyEval_EvalCode ()                                                                                                                              
#8  0x0000000000630af2 in ?? ()                                                                                                                                           
#9  0x0000000000630ba7 in PyRun_FileExFlags ()                                                                                                                            
#10 0x000000000063180f in PyRun_SimpleFileExFlags ()                                                                                                                      
#11 0x0000000000653f7e in ?? ()                                                                                                                                           
#12 0x00000000006542de in _Py_UnixMain ()                                                                                                                                 
#13 0x00007f4d5253209b in __libc_start_main (main=0x4bc7b0 <main>, argc=2, argv=0x7ffc13e5b048, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>,    
    stack_end=0x7ffc13e5b038) at ../csu/libc-start.c:308                                                                                                                  
#14 0x00000000005dfe9a in _start ()                                                                                                                                       
```

### Analysis

If the object passed into :

https://github.com/numpy/numpy/blob/13de4d8f2b803b5775861c825a1ef29b2433f2c5/numpy/core/src/multiarray/descriptor.c#L1366

is a bytes object, then this branch is followed:

https://github.com/numpy/numpy/blob/13de4d8f2b803b5775861c825a1ef29b2433f2c5/numpy/core/src/multiarray/descriptor.c#L1467

This converts the bytes into a c "char *" string:

https://github.com/numpy/numpy/blob/13de4d8f2b803b5775861c825a1ef29b2433f2c5/numpy/core/src/multiarray/descriptor.c#L1472

And then if the length is 1:

https://github.com/numpy/numpy/blob/13de4d8f2b803b5775861c825a1ef29b2433f2c5/numpy/core/src/multiarray/descriptor.c#L1522-L1524

then the first character in the string is taken.  which in this context evaluates to `-1` not `255`.

This value is assumed to be an index into the `_builtin_descrs` array.

When `PyArray_DescrFromType` is then called with this value, there is a check that the value is a valid index for the array before extracting the relevant pointer:

https://github.com/numpy/numpy/blob/13de4d8f2b803b5775861c825a1ef29b2433f2c5/numpy/core/src/multiarray/arraytypes.c.src#L4417-L4419

**BUT**  `-1` < NPY_NTYPES, so the function accepts check_num as valid, despite being out of bounds for the array.

This causes problems when it tries to increase the reference count for the object pointed to by `_builtin_descrs[-1]` on:

https://github.com/numpy/numpy/blob/13de4d8f2b803b5775861c825a1ef29b2433f2c5/numpy/core/src/multiarray/arraytypes.c.src#L4468

Which is not a valid PyObject (and is probably either a canary, other variable, or random address depending on the compiler).

### Numpy/Python version information:

numpy built from HEAD

1.18.0.dev0+13de4d8 3.7.3 (default, Apr  3 2019, 05:39:12)         
[GCC 8.3.0]

**Comment by seberg** [comment]
Thanks, nice tracking down! Do you want to work on a PR to fix this?

**Comment by stestagg** [comment]
Thanks for the response.  
I'll give it a go, the fix seems quite simple, any gotcha's you're aware of?

**Comment by seberg** [comment]
Cool, thanks, can't think of any right away, just that we need tests and that it would be cool if you can try to format your commits the way numpy usually does it.

**Comment by charris** [comment]
Closed by #13907.
